src/Lean/Message.lean

```diff
@@ -78,11 +78,25 @@ inductive MessageData where
     Example: an inspector that tries to find "definitional equality failures" may look for the tag "DefEqFailure". -/
   | tagged            : Name → MessageData → MessageData
   | trace (data : TraceData) (msg : MessageData) (children : Array MessageData)
-  deriving Inhabited
+  /-- Lazy message data production. The `Dymamic` is expected to be `MessageData`, we
+  use this to work around the positivity restriction. -/
+  | ofLazy (f : Option PPContext → IO Dynamic) (hasSyntheticSorry : MetavarContext → Bool)
+  deriving Inhabited, TypeName
 
 namespace MessageData
 
-/-- Determines whether the message contains any content. -/
+def lazy (f : Option PPContext → IO MessageData) (hasSyntheticSorry : MetavarContext → Bool) : MessageData :=
+  .ofLazy (fun ctx => Dynamic.mk <$> f ctx) hasSyntheticSorry
+
+/-- Lazily formatted text with Info annotations -/
+def ofPPFormat (f : PPFormat) : MessageData :=
+  ofLazy (hasSyntheticSorry := f.hasSyntheticSorry) fun ppctx => do
+    let md ← f.pp ppctx
+    return Dynamic.mk (MessageData.ofFormatWithInfos md)
+
+/-- Determines whether the message contains any content.
+This considers lazily generated messages (`.ofLazy`) to be nonempty.
```

Comment:
Why is this the right spec here? What do we use this check for? A quick comment to that effect would be nice.

Reply:
Good question, let’s see. This is only used in 
```
def throwTacticEx (tacticName : Name) (mvarId : MVarId) (msg : MessageData) : MetaM α :=
  if msg.isEmpty then
    throwError "tactic '{tacticName}' failed\n{mvarId}"
  else
    throwError "tactic '{tacticName}' failed, {msg}\n{mvarId}"
```
and almost all calls to `throwTacticEx` has a manifestly non-empty error message. Exceptions are for example
```
       let some mvarIds ← splitTarget? mvarId | Meta.throwTacticEx `split mvarId ""
```
I’m inclined to side-step the issue by adding a proper message to any `Meta.throwTacticEx` invocation. Or, if that is tricky for some reason, change the type to `msg : Option MessageData := none` and use proper types for optional parameters.

Reply:
Did the latter in #3951, so this issue will simply disappear.